idlharness.js

```diff
@@ -444,156 +444,320 @@ IdlArray.prototype.assert_type_is = function(value, type)
      * is fairly elaborate due to the complexity of WebIDL's types, so it's
      * best to look at the grammar to figure out what properties it might have.
      */
+    var ret = this.check_type_is(value, type);
+    if (ret === true)
+    {
+        assert_true(true, "Passed");
+    }
+    else
+    {
+        ret();
+    }
+};
+//@}
+IdlArray.prototype.check_type_is = function(value, type)
+//@{
+{
+    // Helper function that tests whether a value is valid for the given type.
+    // If it is, return true.  If not, return a callback that when called will
+    // execute the proper failing assertion.  This is weird and needs work, but
+    // union types make our lives difficult.
     if (type.idlType == "any")
     {
         // No assertions to make
-        return;
+        return true;
+    }
+
+    if (type.union) {
+        for (var i = 0; i < type.idlType.length; i++) {
+            if (this.check_type_is(value, type.idlType[i]) === true) {
+                return true;
+            }
+        }
+        assert_true(false, "Attribute has value " + format_value(value)
+            + " which doesn't match any of the types in the union");
     }
 
     if (type.nullable && value === null)
     {
         // This is fine
-        return;
+        return true;
     }
 
     if (type.array)
     {
         // TODO: not supported yet
-        return;
+        return true;
     }
 
     if (type.sequence)
     {
-        assert_true(Array.isArray(value), "is not array");
+        if (!Array.isArray(value))
+        {
+            return () => assert_true(Array.isArray(value), "is not array");
+        }
         if (!value.length)
         {
             // Nothing we can do.
-            return;
+            return true;
         }
-        this.assert_type_is(value[0], type.idlType);
-        return;
+        return this.check_type_is(value[0], type.idlType);
     }
 
     if (type.generic === "Promise") {
-        assert_true("then" in value, "Attribute with a Promise type has a then property");
+        if (!("then" in value))
+        {
+            return () => assert_true("then" in value,
+                                     "Attribute with a Promise type has a then property");
+        }
         // TODO: Ideally, we would check on project fulfillment
         // that we get the right type
         // but that would require making the type check async
-        return;
+        return true;
     }
 
     if (type.generic === "FrozenArray") {
-        assert_true(Array.isArray(value), "Value should be array");
-        assert_true(Object.isFrozen(value), "Value should be frozen");
+        if (!Array.isArray(value))
+        {
+            return () => assert_true(Array.isArray(value), "Value should be array");
```

Comment:
I was going to explain how you misunderstood my (admittedly bizarre) code, but then I realized that a much simpler approach should work, so I'm going to scrap this and open a new PR.